src/main/kotlin/com/frcteam3636/frc2025/subsystems/drivetrain/AbsolutePose.kt

```diff
@@ -51,47 +51,111 @@ class AbsolutePoseProviderInputs : LoggableInputs {
     }
 }
 
-
 interface AbsolutePoseProvider {
     fun updateInputs(inputs: AbsolutePoseProviderInputs)
+}
+
+/**
+ * A Limelight localization algorithm.
+ */
+sealed class LimelightAlgorithm {
+    /**
+     * An older, less accurate localization algorithm.
+     */
+    object MegaTag : LimelightAlgorithm()
 
     /**
-     * If there is little to no ambiguity.
+     * A newer and much more accurate algorithm that requires accurate gyro readings and a right-side-up Limelight.
      */
-    val hasHighQualityReading: Boolean
+    class MegaTag2(private val inputsProvider: () -> DrivetrainInputs) : LimelightAlgorithm() {
+        val gyroPosition: Rotation2d
+            get() = inputsProvider().gyroRotation
+        val gyroVelocity: AngularVelocity
+            get() = inputsProvider().gyroVelocity
+    }
 }
 
-class LimelightPoseProvider(private val name: String) : AbsolutePoseProvider {
-    override fun updateInputs(inputs: AbsolutePoseProviderInputs) {
-        // This is mostly pulled from the LimeLight docs.
-        // https://docs.limelightvision.io/docs/docs-limelight/apis/limelight-lib#4-field-localization-with-megatag
-        val estimate = LimelightHelpers.getBotPoseEstimate_wpiBlue(name)
-        val trackedTag = LimelightHelpers.getTargetPose3d_RobotSpace(name)
-
-        // Only trust the measurement if we see multiple tags.
-        inputs.measurement = if (estimate != null && estimate.tagCount >= 2) {
-            AbsolutePoseMeasurement(
-                estimate.pose,
-                Seconds.of(estimate.timestampSeconds),
-                VecBuilder.fill(0.7, 0.7, 9999999.0),
-            )
-        } else {
-            null
+class LimelightPoseProvider(
+    private val name: String,
+    private val algorithm: LimelightAlgorithm,
+) : AbsolutePoseProvider {
+    // References:
+    // https://docs.limelightvision.io/docs/docs-limelight/tutorials/tutorial-swerve-pose-estimation
+    // https://docs.limelightvision.io/docs/docs-limelight/apis/limelight-lib#4-field-localization-with-megatag
+
+    private fun updateCurrentMeasurement(): AbsolutePoseMeasurement? {
+        return when (algorithm) {
+            is LimelightAlgorithm.MegaTag ->
+                LimelightHelpers.getBotPoseEstimate_wpiBlue(name)?.let { estimate ->
+                    // Reject zero tag or low-quality one tag readings
+                    if (estimate.tagCount == 0) return null
```

Comment:
Limelight helpers handles this for you. https://github.com/LimelightVision/limelightlib-wpijava/blob/main/LimelightHelpers.java#L679

Reply:
It looks like that code is handling the case where the Limelight is disconnected, not the case where it is connected and doesn't see any tags. This is also a check that the Limelight sample code does: https://docs.limelightvision.io/docs/docs-limelight/tutorials/tutorial-swerve-pose-estimation

Reply:
The pose is null if no tags are visible.

Reply:
Just leave it in. Doesn't really matter.